lib/spack/spack/relocate_text.py

```diff
@@ -0,0 +1,288 @@
+
+    def apply_to_filename(self, filename):
+        if self.is_noop:
+            return
+        with open(filename, "rb+") as f:
+            self.apply_to_file(f)
+
+    def apply_to_file(self, f):
+        if self.is_noop:
+            return
+        self._apply_to_file(f)
+
+
+class TextFilePrefixReplacer(PrefixReplacer):
+    """This class applies prefix to prefix mappings for relocation
+    on text files.
+
+    Note that UTF-8 encoding is assumed."""
+
+    def __init__(self, prefix_to_prefix: Dict[bytes, bytes]):
+        """
+        prefix_to_prefix (OrderedDict): OrderedDictionary where the keys are
+            bytes representing the old prefixes and the values are the new.
+        """
+        super().__init__(prefix_to_prefix)
+        # Single regex for all paths.
+        self.regex = _byte_strings_to_single_binary_regex(self.prefix_to_prefix.keys())
+
+    @classmethod
+    def from_strings_or_bytes(
+        cls, prefix_to_prefix: Dict[Prefix, Prefix]
+    ) -> "TextFilePrefixReplacer":
+        """Create a TextFilePrefixReplacer from an ordered prefix to prefix map."""
+        return cls(_prefix_to_prefix_as_bytes(prefix_to_prefix))
+
+    def _apply_to_file(self, f):
+        """Text replacement implementation simply reads the entire file
+        in memory and applies the combined regex."""
+        replacement = lambda m: m.group(1) + self.prefix_to_prefix[m.group(2)] + m.group(3)
+        data = f.read()
+        new_data = re.sub(self.regex, replacement, data)
+        if id(data) == id(new_data):
+            return
+        f.seek(0)
+        f.write(new_data)
+        f.truncate()
+
+
+class BinaryFilePrefixReplacer(PrefixReplacer):
+    def __init__(self, prefix_to_prefix, suffix_safety_size=7):
+        """
+        prefix_to_prefix (OrderedDict): OrderedDictionary where the keys are
+            bytes representing the old prefixes and the values are the new
+        suffix_safety_size (int): in case of null terminated strings, what size
+            of the suffix should remain to avoid aliasing issues?
+        """
+        assert suffix_safety_size >= 0
+        super().__init__(prefix_to_prefix)
+        self.suffix_safety_size = suffix_safety_size
+        self.regex = self.binary_text_regex(self.prefix_to_prefix.keys(), suffix_safety_size)
+
+    @classmethod
+    def binary_text_regex(cls, binary_prefixes, suffix_safety_size=7):
+        """
+        Create a regex that looks for exact matches of prefixes, and also tries to
+        match a C-string type null terminator in a small lookahead window.
+
+        Arguments:
+            binary_prefixes (list): List of byte strings of prefixes to match
+            suffix_safety_size (int): Sizeof the lookahed for null-terminated string.
+
+        Returns: compiled regex
+        """
+        return re.compile(
+            b"("
+            + b"|".join(re.escape(p) for p in binary_prefixes)
+            + b")([^\0]{0,%d}\0)?" % suffix_safety_size
```

Comment:
Fwiw, I didn't double check all the regexes - just a couple at random. They're all moved from previous locations, so they should be fine.